src/backend/joanie/core/models/notifications.py

```diff
@@ -0,0 +1,126 @@
+"""
+Declare and configure the models for the notification part
+"""
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from . import User
+from .base import BaseModel
+
+
+class Notification(BaseModel):
+    """
+    Notification model define a notification about an action on a object instance according a
+    subject instance.
+
+    notif_subject: [GenericForeignKey]
+        Defined using notif_subject_ctype [ContentType] and notif_subject_id [uuid]
+        It represents the user's subject that he want to be notified about.
+        It can be a user course wish or a user course enrollment.
+
+    notif_object: [GenericForeignKey]
+        Defined using notif_object_ctype [ContentType] and notif_object_id [uuid]
+        It represents the object that generate the notification according the user's subject
+        It can be a course run, a product or the relation between a product and an course.
+
+    action: [CharField with NOTIF_ACTIONS choices]
+        It represents the action done on the object that generate the notification
+        The action's values are
+            "ADD": the object has been add in a m2m relationship
+            "CREATE": the object has been create
+
+    notif_type: [CharField with NOTIF_TYPES choices]
+        It's the type of the notification
+        It can be an email or a dashboard notification (not implemented yet)
+
+    notified_at: [DateTimeField]
+        It's the DatiTime when the notification has been sent to the owner.
+        If the value is None, the notification has not been sent yet.
+
+    owner: [User]
+        It's the User that received or will receive the notification.
+    """
+
+    NOTIF_ACTION_ADD = "ADD"
+    NOTIF_ACTION_CREATE = "CREATE"
+
+    NOTIF_ACTIONS = (
+        (NOTIF_ACTION_ADD, _("added")),
+        (NOTIF_ACTION_CREATE, _("created")),
+    )
+
+    NOTIF_TYPE_EMAIL = "EMAIL"
+
+    NOTIF_TYPES = (
+        (NOTIF_TYPE_EMAIL, _("Send an email")),
+    )
+
+    notif_subject_ctype_limit = models.Q(app_label='core', model='coursewish') \
+                                | models.Q(app_label='core', model='enrollment')
```

Comment:
In the future, external services should be able to create notifications. The kind of limit you are defining here could block this kind of feature, I think.